internal/cfgmgr/cfgmgr.go

```diff
@@ -0,0 +1,150 @@
+
+//sys cmGetDeviceIDListSize(pulLen *uint32, pszFilter *byte, uFlags uint32) (hr error) = cfgmgr32.CM_Get_Device_ID_List_SizeA
+//sys cmGetDeviceIDList(pszFilter *byte, buffer *byte, bufferLen uint32, uFlags uint32) (hr error)= cfgmgr32.CM_Get_Device_ID_ListA
+//sys cmLocateDevNode(pdnDevInst *uint32, pDeviceID string, uFlags uint32) (hr error) = cfgmgr32.CM_Locate_DevNodeW
+//sys cmGetDevNodeProperty(dnDevInst uint32, propertyKey *devPropKey, propertyType *uint32, propertyBuffer *uint16, propertyBufferSize *uint32, uFlags uint32) (hr error) = cfgmgr32.CM_Get_DevNode_PropertyW
+
+type devPropKey struct {
+	fmtid guid.GUID
+	pid   uint32
+}
+
+const (
+	_CM_GETIDLIST_FILTER_BUSRELATIONS = uint32(0x00000020)
+
+	_CM_LOCATE_DEVNODE_NORMAL = uint32(0x00000000)
+
+	_DEVPROP_TYPE_STRING        = uint32(0x00000012)
+	_DEVPROP_TYPEMOD_LIST       = uint32(0x00002000)
+	_DEVPROP_TYPE_STRING_LIST   = uint32(_DEVPROP_TYPE_STRING | _DEVPROP_TYPEMOD_LIST)
+	_DEVPKEY_LOCATIONPATHS_GUID = "a45c254e-df1c-4efd-8020-67d146a850e0"
+)
+
+// getDevPKeyDeviceLocationPaths creates a DEVPROPKEY struct for the
+// DEVPKEY_Device_LocationPaths property as defined in devpkey.h
+func getDevPKeyDeviceLocationPaths() (*devPropKey, error) {
+	guid, err := guid.FromString(_DEVPKEY_LOCATIONPATHS_GUID)
+	if err != nil {
+		return nil, err
+	}
+	return &devPropKey{
+		fmtid: guid,
+		// pid value is defined in devpkey.h
+		pid: 37,
+	}, nil
+}
+
+func GetDeviceLocationPathsFromIDs(ids []string) ([]string, error) {
+	result := []string{}
+	devPKeyDeviceLocationPaths, err := getDevPKeyDeviceLocationPaths()
+	if err != nil {
+		return nil, err
+	}
+	for _, id := range ids {
+		var devNodeInst uint32
+		err = cmLocateDevNode(&devNodeInst, id, _CM_LOCATE_DEVNODE_NORMAL)
+		if err != nil {
+			return nil, errors.Wrapf(err, "failed to locate device node for %s", id)
+		}
+		propertyType := uint32(0)
+		propertyBufferSize := uint32(0)
+
+		// get the size of the property buffer by querying with a nil buffer and zeroed propertyBufferSize
+		err = cmGetDevNodeProperty(devNodeInst, devPKeyDeviceLocationPaths, &propertyType, nil, &propertyBufferSize, 0)
+		if err != nil {
+			return nil, errors.Wrapf(err, "failed to get property buffer size of devnode query for %s with", id)
+		}
+
+		// get the property with the resulting propertyBufferSize
+		propertyBuffer := make([]uint16, propertyBufferSize/2)
+		err = cmGetDevNodeProperty(devNodeInst, devPKeyDeviceLocationPaths, &propertyType, &propertyBuffer[0], &propertyBufferSize, 0)
+		if err != nil {
+			return nil, errors.Wrapf(err, "failed to get location path property from device node for %s with", id)
+		}
+		if propertyType != _DEVPROP_TYPE_STRING_LIST {
+			return nil, fmt.Errorf("expected to return property type DEVPROP_TYPE_STRING_LIST %d, instead got %d", _DEVPROP_TYPE_STRING_LIST, propertyType)
+		}
+		if int(propertyBufferSize/2) > len(propertyBuffer) {
```

Comment:
I'm hesitant to do that since I think it's possible the the value could have changed between the two calls, but it may not necessarily be an error if the contents of the buffer is smaller than previously expected.